### PR TITLE
feat(template): add widget template

### PR DESCRIPTION
**Pull request by Haroenv**
this changes the questions asked into two categories, and asks one extra question for the integration type.

I had to split up getConfiguration, since I need to know which template category is used to decide around the further questions. 

This is in reference to DX-1605

### CODE REVIEW

File: src/cli/index.js [hunk 1]
```diff
@@ -90,133 +94,195 @@ try {
...
-    },
-    validate(input) {
-      return Boolean(input);
+const questions = {
+  application: [
+    {
+      type: 'list',
+      pageSize: 10,
+      name: 'template',
+      message: 'InstantSearch template',
```
Comment (Haroenv) [inline src/cli/index.js:103]: if this question comes first, we can make only the questions after this conditional, and no longer need the new "implementation type" question

File: src/cli/index.js [hunk 2]
```diff
@@ -90,140 +94,204 @@ try {
...
+      filter: attributes => attributes.filter(Boolean),
+      when: ({ appId, apiKey, indexName }) =>
+        !attributesToDisplay.length > 0 && appId && apiKey && indexName,
+    },
+  ],
+  widget: [
+    {
+      type: 'input',
+      name: 'organization',
+      message: 'organization (on npm)',
```
Comment (francoischalifour) [inline src/cli/index.js:196]:
```suggestion
      message: 'Organization (on npm)',
```

File: src/cli/index.js [hunk 3]
```diff
@@ -90,140 +94,204 @@ try {
...
-    filter: attributes => attributes.filter(Boolean),
-    when: ({ appId, apiKey, indexName }) =>
-      !attributesToDisplay.length > 0 && appId && apiKey && indexName,
-  },
-].filter(question => isQuestionAsked({ question, args: optionsFromArguments }));
+    },
+    {
+      type: 'input',
+      name: 'description',
+      message: 'npm and readme description',
```
Comment (francoischalifour) [inline src/cli/index.js:204]:
```suggestion
      message: 'Description',
```

File: src/templates/InstantSearch.js widget/.template.js [hunk 4]
```diff
@@ -0,0 +1,15 @@
+const install = require('../../tasks/node/install');
+const teardown = require('../../tasks/node/teardown');
+
+module.exports = {
+  category: 'Web - Widget',
```
Comment (francoischalifour) [inline src/templates/InstantSearch.js widget/.template.js:5]:
```suggestion
  category: 'Widget',
```

File: src/cli/index.js [hunk 5]
```diff
@@ -90,140 +94,204 @@ try {
...
+      attributesToDisplay,
+    },
+    answers: { template },
+  });
+
+  const templatePath = getTemplatePath(configuration.template);
+  const templateConfig = getAppTemplateConfig(templatePath);
+
+  const implementationType =
+    templateConfig.category === 'Web - Widget' ? 'widget' : 'application';
```
Comment (francoischalifour) [inline src/cli/index.js:264]:
```suggestion
    templateConfig.category === 'Widget' ? 'widget' : 'application';
```

File: src/cli/__tests__/getConfiguration.test.js [hunk 6]
```diff
@@ -18,18 +17,6 @@ test('without template throws', async () => {
   );
 });
 
-test('with template transforms to its relative path', async () => {
```
Comment (francoischalifour) [inline src/cli/__tests__/getConfiguration.test.js:21]: The generation of other apps breaks on my machine now.

File: src/templates/InstantSearch.js widget/README.md [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+search.start();
+```
+
+### Options
+
+#### container
+
+type: string | Element
+
+required: true
```
[the remark this answers is not in the document]
Comment (Haroenv) [reply src/templates/InstantSearch.js widget/README.md:41]: 22a7716

File: src/templates/InstantSearch.js widget/README.md [hunk 8]
```diff
@@ -0,0 +1,159 @@
...
+## Widget
+
+### Usage
+
+```js
+import { {{ camelCaseName }} } from '{{ packageName }}';
+
+const search = instantsearch({
+  indexName: 'indexName',
+  searchClient: algoliasearch('appId', 'apiKey'),
```
[the remark this answers is not in the document]
Comment (Haroenv) [reply src/templates/InstantSearch.js widget/README.md:23]: 22a7716

File: src/cli/__tests__/getConfiguration.test.js (hunk 6 above)
Comment (Haroenv) [reply src/cli/__tests__/getConfiguration.test.js:21]: fixed in 48c3905

**Review by francoischalifour — COMMENTED** [review]
I still get this error:

```
$ node src/cli/index.js /tmp/hello-widget

Creating a new InstantSearch app in /tmp/hello-widget.

? InstantSearch template InstantSearch.js
The template configuration file `.template.js` contains errors:
The "path" argument must be of type string. Received type undefined
```

**Review by francoischalifour — APPROVED** [review]
Feel free to merge this once the last comments are addressed!

File: src/templates/InstantSearch.js widget/src/connector.ts.hbs [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+            instantSearchInstance,
+          },
+          false
+        );
+      },
+
+      /*
+       * Called when this widget is unmounted.
+       * Can be used to remove refinements set by during this widget's initialization and life time.
+       * See `dispose` documentation https://www.algolia.com/doc/
```
[the remark this answers is not in the document]
Comment (Haroenv) [reply src/templates/InstantSearch.js widget/src/connector.ts.hbs:62]: we don't actually have proper doc on custom widgets at the moment, so I will remove the "see xxx documentation" for now

**Pull request merged**